minigui/graph.ts

```diff
@@ -0,0 +1,190 @@
+import {getElement} from './util'
+
+type MoveChangedCallback = (move: number | null) => void;
+
+class Graph {
+  protected ctx: CanvasRenderingContext2D;
+  protected points = new Array<[number, number]>();
+  protected pixelRatio = window.devicePixelRatio || 1;
+  protected marginTop = 20 * this.pixelRatio;
+  protected marginBottom = 20 * this.pixelRatio;
+  protected marginLeft = 30 * this.pixelRatio;
+  protected marginRight = 60 * this.pixelRatio;
+  protected minPoints = 10;
+
+  protected w: number;
+  protected h: number;
+  protected selectedMove: number | null = null;
+  protected moveChangedCallback: MoveChangedCallback | null = null;
+
+  constructor(parent: HTMLElement | string) {
+    if (typeof(parent) == 'string') {
+      parent = getElement(parent);
+    }
+
+    let canvas = document.createElement('canvas');
+    canvas.width = this.pixelRatio * parent.offsetWidth;
+    canvas.height = this.pixelRatio * parent.offsetHeight;
+    canvas.style.width = `${parent.offsetWidth}px`;
+    canvas.style.height = `${parent.offsetHeight}px`;
+    this.ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
+
+    this.w = canvas.width - this.marginLeft - this.marginRight;
+    this.h = canvas.height - this.marginTop - this.marginBottom;
+
+    parent.appendChild(canvas);
+
+    canvas.addEventListener('mousemove', (e) => {
+      this.onMouseMove(e.offsetX, e.offsetY);
+    });
+    canvas.addEventListener('mouseleave', () => {
+      this.onMouseLeave();
+    });
+
+    this.draw();
+  }
+
+  clear() {
+    this.points = [];
+    this.draw();
+  }
+
+  setPoint(x: number, y: number) {
```

Comment:
this probably needs a comment; looks like this is actually more like `addPoint`, semantically, but with some magic around not overscaling the opening moves?

Reply:
added a comment and made the logic less obscure